### DISCUSSION TITLE
Doesn't work in crontab?

### DISCUSSION THREAD

**Discussion by lucky3582** [BurntSushi/ripgrep#1979] [Q&A] [answered]
I am running a cronjob which calls a python script which has subprocess checkoutput.

I tried a simple script to debug.

```
import subprocess
import os

command = "/home/sgadamse/history_checker/code/rg \"hello\""
try:
    result = subprocess.check_output(command,shell=True)
    print(result)
except subprocess.CalledProcessError as e:
    print(e)
```

Output when ran in the shell: I get the output i needed
```

[~/history_checker/code]$ python a.py                                                                                                                                 

a.py:command = "/home/sgadamse/history_checker/code/rg \"hello\""
```

I created a cron job to run in crontab -e

`* * * * * python /home/sgadamse/history_checker/code/a.py`
I get this error when its executed:

`Command '/home/sgadamse/history_checker/code/rg "hello"' returned non-zero exit status 1`

/home/sgadamse/history_checker/code/rg is the complete rg binary path that i downloaded and using it.

Do we need to do anything differently for the the cron jobs to execute the binary?

Thanks.

Edit:
I added --stats option its searching only one file i.e <stdin> in the cron job 

Any idea why its not searching only stdin? 
i also added to cwd to the python command but its still the same

Edit2:

I tried the grep command with exact cron job and python script and works as expected.
Am i missing anything for rg? :(
I also tried to use rg but installing it directly instead of specifying the binary name. Still the same

**Comment by BurntSushi** [comment, marked as answer]
From the [Python docs for subprocess](https://docs.python.org/3/library/subprocess.html#subprocess.Popen), emphasis mine:

> stdin, stdout and stderr specify the executed program’s standard input, standard output and standard error file handles, respectively. Valid values are PIPE, DEVNULL, an existing file descriptor (a positive integer), an existing file object, and None. PIPE indicates that a new pipe to the child should be created. DEVNULL indicates that the special file os.devnull will be used. **With the default settings of None, no redirection will occur; the child’s file handles will be inherited from the parent.**

My guess is that your cron system is attaching something to stdin that is being inherited by your Python process. Try using [`stdin=subprocess.DEVNULL`](https://docs.python.org/3/library/subprocess.html#subprocess.DEVNULL).

Alternatively, tell ripgrep to search a specific thing instead of having to guess. e.g., `rg hello ./` instead of `rg hello`. With the latter, ripgrep has to guess as to whether to search the current directory or to search stdin. Consider how ripgrep knows what to do in both `rg hello` and `echo hello | rg hello`. The commands are the same, but their behavior is different due to how stdin is being used.

**Reply by lucky3582** [reply]
Thanks for the reply. I tried the ./ and it worked!
